stablehlo/reference/Ops.cpp

```diff
@@ -37,6 +38,194 @@ Index evalIndices(ArrayRef<Tensor> runtimeIndices) {
+  return RankedTensorType::get(inferredConvolutionType[0].getDims(),
+                               resultType.getElementType());
+}
+
+TensorType inferDotGeneralOpType(TensorType lhsType, TensorType rhsType,
+                                 ArrayRef<int64_t> lhsContractingDimensions,
+                                 ArrayRef<int64_t> rhsContractingDimensions) {
+  SmallVector<ShapedTypeComponents> inferredDotGeneralType;
+  auto dotGeneralStatus = hlo::inferDotGeneralOp(
+      /*location=*/{}, lhsType, rhsType,
+      /*lhsBatchingDimensions=*/{}, /*rhsBatchingDimensions*/ {},
+      lhsContractingDimensions, rhsContractingDimensions,
+      /*precisionConfig=*/{}, inferredDotGeneralType);
+  if (failed(dotGeneralStatus))
+    report_fatal_error(
+        invalidArgument("Could not infer DotGeneralOp's return type"));
+  return RankedTensorType::get(inferredDotGeneralType[0].getDims(),
+                               lhsType.getElementType());
+}
+
+TensorType inferPadOpType(ArrayRef<std::pair<int64_t, int64_t>> lhsPadding,
+                          Type operandType, Type resultElementType,
+                          ArrayRef<int64_t> interiorPadding) {
+  Builder builder = mlir::Builder(operandType.getContext());
+  Type i64Type = builder.getI64Type();
+  SmallVector<int64_t> lhsPaddingLow;
+  SmallVector<int64_t> lhsPaddingHigh;
+  for (auto paddingPair : lhsPadding) {
+    lhsPaddingLow.push_back(paddingPair.first);
+    lhsPaddingHigh.push_back(paddingPair.second);
+  }
+  SmallVector<Type> inferredPadType;
+  auto padStatus = hlo::inferPadOp(
+      {}, operandType, RankedTensorType::get({}, resultElementType),
+      getDenseIntElementsAttr(i64Type, lhsPaddingLow, {}),
+      getDenseIntElementsAttr(i64Type, lhsPaddingHigh, {}),
+      getDenseIntElementsAttr(i64Type, interiorPadding, {}), inferredPadType);
+  if (failed(padStatus))
+    report_fatal_error(invalidArgument("Could not infer PadOp's return type"));
+  return inferredPadType[0].cast<TensorType>();
+}
+
+TensorType inferSliceOpType(Type operandType,
+                            SmallVector<int64_t> lhsWindowStart,
+                            SmallVector<int64_t> limitIndices,
+                            SmallVector<int64_t> lhsWindowDilations) {
+  Builder builder = mlir::Builder(operandType.getContext());
+  Type i64Type = builder.getI64Type();
+  SmallVector<Type> inferredSliceType;
+  auto sliceStatus = hlo::inferSliceOp(
+      {}, operandType, getDenseIntElementsAttr(i64Type, lhsWindowStart, {}),
+      getDenseIntElementsAttr(i64Type, limitIndices, {}),
+      getDenseIntElementsAttr(i64Type, lhsWindowDilations, {}),
+      inferredSliceType);
+  if (failed(sliceStatus))
+    report_fatal_error(
+        invalidArgument("Could not infer SliceOp's return type"));
+  return inferredSliceType[0].cast<TensorType>();
+}
+
+template <typename T>
+SmallVector<T> concatAndPermute(T n, SmallVector<T> hw, T c, int64_t dimA,
```

Comment:
Thanks for the suggestion!

4. Sizes class has the `operator+`, but this one is adding values to it, not concatenating. It would be better to avoid non-consistent use of the `operator+` between `Axes` and `Sizes` classes.